File: skript-parser-java-impl/src/main/java/io/github/syst3ms/skriptparser/registration/SyntaxManager.java
package io.github.syst3ms.skriptparser.registration;

import io.github.syst3ms.skriptparser.lang.interfaces.DynamicNumberExpression;
import io.github.syst3ms.skriptparser.lang.Effect;
import io.github.syst3ms.skriptparser.lang.Expression;
import io.github.syst3ms.skriptparser.parsing.SkriptParserException;
import io.github.syst3ms.skriptparser.util.MultiMap;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public class SyntaxManager {
    private static SyntaxManager instance = new SyntaxManager();
    private static MultiMap<Class<?>, ExpressionInfo<?, ?>> expressions = new MultiMap<>();
    private static MultiMap<String, SyntaxInfo<?>> syntaxes = new MultiMap<>();
    private static List<SyntaxInfo<? extends Effect>> effects = new ArrayList<>();

    private SyntaxManager() {
    }

    public static SyntaxManager getInstance() {
        return instance;
    }

    public static void register(SkriptRegistration reg) {
        for (SyntaxInfo<? extends Effect> info : reg.getEffects()) {
            effects.add(info);
            syntaxes.putOne(reg.getRegisterer(), info);
        }
        for (Map.Entry<Class<?>, List<ExpressionInfo<?, ?>>> entry : reg.getExpressions().entrySet()) {
            Class<?> key = entry.getKey();
            List<ExpressionInfo<?, ?>> infos = entry.getValue();
            for (ExpressionInfo<?, ?> info : infos) {
                expressions.putOne(key, info);
                syntaxes.putOne(reg.getRegisterer(), info);
            }
        }
    }

    public static Iterable<SyntaxInfo<?>> getAddonSyntaxes(String name) {
        List<SyntaxInfo<?>> infos = syntaxes.get(name);
        return infos == null ? Collections.emptyList() : infos;
    }

    public static Collection<ExpressionInfo<?, ?>> getAllExpressions() {
        return expressions.getAllValues();
    }

    public static  <T> List<ExpressionInfo<?, ?>> getExpressionsByReturnType(Class<? extends T> c) {
        List<ExpressionInfo<?, ?>> infos = new ArrayList<>();
        for (Class<?> returnType : expressions.keySet()) {
            if (returnType.isAssignableFrom(c)) {
                infos.addAll(expressions.get(returnType));
            }
        }
        return infos;
    }

    public static Collection<SyntaxInfo<? extends Effect>> getEffects() {
        return effects;
    }
}

File: skript-parser-kotlin/src/main/kotlin/io/github/syst3ms/skriptparser/registration/SyntaxManager.kt
package io.github.syst3ms.skriptparser.registration

import io.github.syst3ms.skriptparser.util.MultiMap

object SyntaxManager {
    private val expressions = MultiMap<Class<*>, ExpressionInfo<*, *>>()
    private val syntaxes = MultiMap<String, SyntaxInfo<*>>()
    private val effects = arrayListOf<SyntaxInfo<*>>()

    val allExpressions: Iterable<ExpressionInfo<*, *>>
        get() = expressions.allValues

    internal fun register(reg: SkriptRegistration) {
        for (info in reg.effects) {
            effects.add(info)
            syntaxes.putOne(reg.registerer, info)
        }
        for ((key, infos) in reg.expressions) {
            for (info in infos) {
                expressions.putOne(key, info)
                syntaxes.putOne(reg.registerer, info)
            }
        }
    }

    fun getAddonSyntaxes(name: String) = syntaxes[name] ?: emptyList<SyntaxInfo<*>>()

    fun <T> getExpressionsByReturnType(c: Class<out T>): List<ExpressionInfo<*, *>> {
        val infos = arrayListOf<ExpressionInfo<*, *>>()
        for (returnType in expressions.keys) {
            if (returnType.isAssignableFrom(c)) {
                infos.addAll(expressions[returnType]!!)
            }
        }
        return infos
    }
}
